pkg/kinflate/resource/configmap.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	manifest "k8s.io/kubectl/pkg/apis/manifest/v1alpha1"
+	cutil "k8s.io/kubectl/pkg/kinflate/configmapandsecret/util"
+	"k8s.io/kubectl/pkg/kinflate/hash"
+)
+
+// configMap implements Resource interface for ConfigMap type.
+type configMap struct {
+	cm manifest.ConfigMap
+	// add loader here
+
+	name string
+	data *unstructured.Unstructured
+}
+
+// (Note): pass in loader which has rootPath context and knows how to load
+// files given a relative path.
+func NewConfigMap(cm manifest.ConfigMap) Resource {
+	return &configMap{
+		cm: cm,
+	}
+}
+
+func (cm *configMap) Name() string {
+	return cm.name
+}
+
+func (cm *configMap) Data() (unstructured.Unstructured, error) {
+	if cm.data != nil {
+		return *cm.data, nil
+	}
+	d, err := cm.populateData()
+	if err == nil {
+		cm.data = *d
+		return *cm.data, nil
+	}
+	return nil, err
+}
+
+func (cm *configMap) populateData() (*unstructured.Unstructured, error) {
+	corev1CM, err := makeConfigMap(cm)
+	if err != nil {
+		return nil, err
+	}
+	h, err := hash.ConfigMapHash(corev1CM)
+	if err != nil {
+		return nil, err
+	}
+	cm.name = fmt.Sprintf("%s-%s", corev1CM.GetName(), h)
+	return objectToUnstructured(corev1CM)
+}
+
+func makeConfigMap(cm manifest.ConfigMap) (*corev1.ConfigMap, error) {
+	corev1cm := &corev1.ConfigMap{
+		APIVersion: "v1",
+		Kind:       "ConfigMap",
+		Name:       cm.Name,
+		Data:       map[string]string{},
+	}
+
+	if cm.EnvSource != "" {
+		if err := cutil.HandleConfigMapFromEnvFileSource(corev1cm, cm.EnvSource); err != nil {
```

Comment:
do we want to move that cutil code closer in this resource package?

Reply:
Yes, we do.